ddl/ddl_api.go

```diff
@@ -4720,6 +4745,46 @@ func (d *ddl) AlterTableSetTiFlashReplica(ctx sessionctx.Context, ident ast.Iden
 		return errors.Trace(err)
 	}
 
+	// We should check this first, in order to avoid creating redundant DDL jobs.
+	tblInfo := tb.Meta()
+	if d.IsTiFlashPollEnabled() {
+		// Must set placement rule before the ActionSetTiFlashReplica job is in queue.
+		tikvStore, ok := ctx.GetStore().(helper.Storage)
+		if !ok {
+			logutil.BgLogger().Error("can not get Helper")
+		}
+		tikvHelper := &helper.Helper{
+			Store:       tikvStore,
+			RegionCache: tikvStore.GetRegionCache(),
+		}
+		if pi := tblInfo.GetPartitionInfo(); pi != nil {
+			for _, p := range pi.Definitions {
+				logutil.BgLogger().Info("Set TiFlash replica pd rule", zap.Int64("partitionID", p.ID))
+				ruleNew := MakeNewRule(p.ID, replicaInfo.Count, replicaInfo.Labels)
+				if e := tikvHelper.SetPlacementRule(*ruleNew); e != nil {
+					return errors.Trace(err)
+				}
+			}
+			// Partitions that in adding mid-state. They have high priorities, so we should set accordingly pd rules.
+			for _, p := range pi.AddingDefinitions {
+				logutil.BgLogger().Info("Set TiFlash replica pd rule", zap.Int64("addPartitionID", p.ID))
+				ruleNew := MakeNewRule(p.ID, replicaInfo.Count, replicaInfo.Labels)
+				if e := tikvHelper.SetPlacementRule(*ruleNew); e != nil {
```

Comment:
If `replicaInfo.Count == 0` should we `DeletePlacementRule` instead of `SetPlacementRule `

Reply:
Not sure about this, I will do write some test to check that.